// hollidays/src/index.js
import jsonHollidays from './data/data.json'
import Generator from './generator'

class App {

        constructor () {
            this.country = null;
            this.state = null;
        }
        /**
         * @return {Array} - return array of availible countries
         */
        getCountries() {
            return Object.keys(jsonHollidays);
        }

        /**
         * @return {Array} - return array of availible states
         */
        getStates() {
            if (this.country === null) {
                // console.error("Country needs to be set first");
                return false;
            } else if (!jsonHollidays[this.country].hasOwnProperty('states')) {
                // console.error("This country has no states")
                return false;
            }
            return Object.keys(jsonHollidays[this.country].states)
        }

        /**
         * Set country and state and return changed object
         * @param {String} country 
         * @param {String} state 
         * @return {Object}
         */
        setCountry(country, state) {
            if (typeof country != "string" || !jsonHollidays.hasOwnProperty(country)) {
                return false;
            }
            if (typeof state == "string") {
                this.state = state
            }

            this.country = country;
            return this;
        }

        /**
         * Return all holidays in selected country
         * @param {Number} year 
         * @return {Object}
         */
        findAll(year) {
            if (isNaN(year)) {
                // console.error("Parameter must be number");
                return false;
            }

            return Generator(this.country, year, this.state)
        }

        /**
         * Search for holiday on selected date
         * Return object if found or false if not found
         * @param {Number} year 
         * @param {Number} month 
         * @param {Number} day
         * @return {Object | Boolean} 
         */
        findByDate(year, month, day) {
            if (this.country === null) {
                // console.error("Country needs to be set first")
                return false
            }

            month = Number(month).toString();
            day = Number(day).toString();

            let allDates = Generator(this.country, year, this.state);
            
            let result = allDates.find(holliday => {
                return holliday.date == day + '.' + month;
            })

            if(typeof result === "undefined"){
                return false;
            }
            return result;
        }
}

export default new App();
